File: scripts/t-wrapper/ast-transformers.ts
/**
 * AST 변환 로직
 * 문자열 리터럴, 템플릿 리터럴, JSX 텍스트를 t() 함수로 변환
 */

import { NodePath } from "@babel/traverse";
import * as t from "@babel/types";
import { hasIgnoreComment, shouldSkipPath } from "./ast-helpers";
import { STRING_CONSTANTS, REGEX_PATTERNS } from "./constants";

export interface TransformResult {
  wasModified: boolean;
}

/**
 * 함수 body 내의 AST 노드들을 변환
 */
export function transformFunctionBody(
  path: NodePath<t.Function>,
  sourceCode: string
): TransformResult {
  let wasModified = false;

  path.traverse({
    StringLiteral: (subPath) => {
      if (
        shouldSkipPath(subPath, hasIgnoreComment) ||
        hasIgnoreComment(subPath, sourceCode)
      ) {
        return;
      }

      // 빈 문자열이나 공백만 있는 문자열은 스킵
      const trimmedValue = subPath.node.value.trim();
      if (!trimmedValue) {
        return;
      }

      // 한국어 텍스트가 포함된 문자열만 처리
      if (REGEX_PATTERNS.KOREAN_TEXT.test(subPath.node.value)) {
        wasModified = true;
        const replacement = t.callExpression(
          t.identifier(STRING_CONSTANTS.TRANSLATION_FUNCTION),
          [t.stringLiteral(subPath.node.value)]
        );

        if (t.isJSXAttribute(subPath.parent)) {
          subPath.replaceWith(t.jsxExpressionContainer(replacement));
        } else {
          subPath.replaceWith(replacement);
        }
      }
    },
    TemplateLiteral: (subPath) => {
      // i18n-ignore 주석이 있는 경우 스킵
      if (
        shouldSkipPath(subPath as any, hasIgnoreComment) ||
        hasIgnoreComment(subPath, sourceCode)
      ) {
        return;
      }

      // 이미 t()로 래핑된 경우 스킵
      if (
        t.isCallExpression(subPath.parent) &&
        t.isIdentifier(subPath.parent.callee, {
          name: STRING_CONSTANTS.TRANSLATION_FUNCTION,
        })
      ) {
        return;
      }

      // 템플릿 리터럴의 모든 부분에 하나라도 한국어가 있는지 확인
      const hasKorean = subPath.node.quasis.some((quasi) =>
        REGEX_PATTERNS.KOREAN_TEXT.test(quasi.value.raw)
      );

      if (!hasKorean) {
        return;
      }

      // 템플릿 리터럴을 i18next interpolation 형식으로 변환
      // 예: `안녕 ${name}` → t(`안녕 {{name}}`, { name })
      wasModified = true;

      const templateNode = subPath.node;
      const expressions = templateNode.expressions;
      const quasis = templateNode.quasis;

      // 표현식이 없으면 단순 문자열로 처리
      if (expressions.length === 0) {
        const replacement = t.callExpression(
          t.identifier(STRING_CONSTANTS.TRANSLATION_FUNCTION),
          [t.stringLiteral(quasis[0].value.raw)]
        );
        subPath.replaceWith(replacement);
        return;
      }

      // i18next 형식으로 변환: `안녕 ${name}` → `안녕 {{name}}`
      let i18nextString = "";
      const interpolationVars: t.ObjectProperty[] = [];

      quasis.forEach((quasi, index) => {
        i18nextString += quasi.value.raw;

        if (index < expressions.length) {
          const expr = expressions[index];

          // 변수명 추출
          let varName: string;
          if (t.isIdentifier(expr)) {
            varName = expr.name;
          } else if (t.isMemberExpression(expr)) {
            // user.name → user_name
            // 간단한 멤버 표현식은 직접 변환
            const parts: string[] = [];
            let current: any = expr;
            while (t.isMemberExpression(current)) {
              if (t.isIdentifier(current.property)) {
                parts.unshift(current.property.name);
              }
              current = current.object;
            }
            if (t.isIdentifier(current)) {
              parts.unshift(current.name);
            }
            varName = parts.join(STRING_CONSTANTS.MEMBER_SEPARATOR);
          } else {
            // 복잡한 표현식은 expr0, expr1 등으로 처리
            varName = `${STRING_CONSTANTS.EXPR_PREFIX}${index}`;
          }

          // i18next 형식: {{varName}}
          i18nextString += `${STRING_CONSTANTS.INTERPOLATION_START}${varName}${STRING_CONSTANTS.INTERPOLATION_END}`;

          // interpolation 객체에 추가
          interpolationVars.push(
            t.objectProperty(t.identifier(varName), expr as t.Expression)
          );
        }
      });

      // t("안녕 {{name}}", { name: name })
      const args: Array<t.Expression | t.SpreadElement> = [
        t.stringLiteral(i18nextString),
      ];

      // interpolation 객체가 있으면 두 번째 인자로 추가
      if (interpolationVars.length > 0) {
        args.push(t.objectExpression(interpolationVars));
      }

      const replacement = t.callExpression(
        t.identifier(STRING_CONSTANTS.TRANSLATION_FUNCTION),
        args
      );
      subPath.replaceWith(replacement);
    },
    JSXText: (subPath) => {
      // i18n-ignore 주석이 있는 경우 스킵
      if (hasIgnoreComment(subPath, sourceCode)) {
        return;
      }

      const text = subPath.node.value.trim();

      // 빈 텍스트나 공백만 있는 경우 스킵
      if (!text) {
        return;
      }

      // 한국어가 포함된 텍스트만 처리
      if (REGEX_PATTERNS.KOREAN_TEXT.test(text)) {
        wasModified = true;

        // t() 함수 호출로 감싸기
        const replacement = t.jsxExpressionContainer(
          t.callExpression(t.identifier(STRING_CONSTANTS.TRANSLATION_FUNCTION), [
            t.stringLiteral(text),
          ])
        );

        subPath.replaceWith(replacement);
      }
    },
  });

  return { wasModified };
}

File: scripts/t-wrapper-rust/src/ast_transformers.rs
/// AST 변환 로직
/// 문자열 리터럴, 템플릿 리터럴, JSX 텍스트를 t() 함수로 변환

use crate::constants::{StringConstants, RegexPatterns};
use crate::ast_helpers::{has_ignore_comment, should_skip_path};
use swc_ecma_ast::*;
use swc_ecma_visit::{VisitMut, VisitMutWith};
use swc_common::DUMMY_SP;

/// 변환 결과
#[derive(Debug, Clone)]
pub struct TransformResult {
    pub was_modified: bool,
}

impl TransformResult {
    pub fn new(was_modified: bool) -> Self {
        Self { was_modified }
    }
}

/// 함수 body 내의 AST 노드들을 변환
/// 
/// TypeScript 버전과 동일한 로직:
/// 1. StringLiteral: 한국어가 포함된 문자열을 t() 호출로 변환
/// 2. TemplateLiteral: 템플릿 리터럴을 i18next 형식으로 변환
/// 3. JSXText: JSX 텍스트를 t() 호출로 변환
/// 
/// TODO: SWC AST traverse로 구현 필요
/// 현재는 소스코드에서 한국어 감지만 수행
pub fn transform_function_body(_path: (), source_code: &str) -> TransformResult {
    let mut was_modified = false;

    // TODO: SWC AST traverse로 구현
    // path.traverse({
    //     StringLiteral: (subPath) => {
    //         if (shouldSkipPath(subPath, hasIgnoreComment) || hasIgnoreComment(subPath, sourceCode)) {
    //             return;
    //         }
    //         const trimmedValue = subPath.node.value.trim();
    //         if (!trimmedValue) {
    //             return;
    //         }
    //         if (REGEX_PATTERNS.KOREAN_TEXT.test(subPath.node.value)) {
    //             wasModified = true;
    //             const replacement = t.callExpression(
    //                 t.identifier(STRING_CONSTANTS.TRANSLATION_FUNCTION),
    //                 [t.stringLiteral(subPath.node.value)]
    //             );
    //             if (t.isJSXAttribute(subPath.parent)) {
    //                 subPath.replaceWith(t.jsxExpressionContainer(replacement));
    //             } else {
    //                 subPath.replaceWith(replacement);
    //             }
    //         }
    //     },
    //     TemplateLiteral: (subPath) => {
    //         // i18n-ignore 주석이 있는 경우 스킵
    //         // 이미 t()로 래핑된 경우 스킵
    //         // 템플릿 리터럴을 i18next interpolation 형식으로 변환
    //         // 예: `안녕 ${name}` → t(`안녕 {{name}}`, { name })
    //     },
    //     JSXText: (subPath) => {
    //         // i18n-ignore 주석이 있는 경우 스킵
    //         // 한국어가 포함된 텍스트만 처리
    //         // t() 함수 호출로 감싸기
    //     },
    // });

    // 임시로 한국어가 포함되어 있으면 수정되었다고 가정
    if RegexPatterns::korean_text().is_match(source_code) {
        was_modified = true;
    }

    TransformResult::new(was_modified)
}

/// SWC AST Module을 변환하는 Transformer
pub struct TranslationTransformer {
    pub was_modified: bool,
    source_code: String,
}

impl TranslationTransformer {
    pub fn new(source_code: String) -> Self {
        Self {
            was_modified: false,
            source_code,
        }
    }

    /// t() 함수 호출 생성
    /// TODO: 실제 AST 노드 생성 구현
    fn create_t_call(&self, _value: &str) -> Expr {
        // TODO: 실제 AST 노드 생성 구현
        // 현재는 플레이스홀더
        Expr::Ident(Ident {
            span: DUMMY_SP,
            sym: StringConstants::TRANSLATION_FUNCTION.into(),
            optional: false,
            ctxt: Default::default(),
        })
    }
}

impl VisitMut for TranslationTransformer {
    /// Expression 변환 (StringLiteral을 t() 호출로 교체)
    /// TypeScript 버전과 동일한 로직:
    /// 1. StringLiteral 감지
    /// 2. 한국어 텍스트가 포함된 문자열만 처리
    /// 3. t() 함수 호출로 변환
    fn visit_mut_expr(&mut self, expr: &mut Expr) {
        // StringLiteral을 t() 호출로 교체
        if let Expr::Lit(Lit::Str(str_lit)) = expr {
            // Wtf8Atom을 문자열로 변환
            let value_str = str_lit.value.to_string();
            
            // 빈 문자열이나 공백만 있는 문자열은 스킵
            let trimmed = value_str.trim();
            if trimmed.is_empty() {
                return;
            }
            
            // 한국어 텍스트가 포함된 문자열만 처리
            if RegexPatterns::korean_text().is_match(&value_str) {
                self.was_modified = true;
                
                // t() 함수 호출 생성
                let t_call = Expr::Call(CallExpr {
                    span: DUMMY_SP,
                    callee: Callee::Expr(Box::new(Expr::Ident(Ident {
                        span: DUMMY_SP,
                        sym: StringConstants::TRANSLATION_FUNCTION.into(),
                        optional: false,
                        ctxt: Default::default(),
                    }))),
                    args: vec![ExprOrSpread {
                        spread: None,
                        expr: Box::new(Expr::Lit(Lit::Str(Str {
                            span: str_lit.span,
                            value: str_lit.value.clone(),
                            raw: None,
                        }))),
                    }],
                    type_args: None,
                });
                
                // 현재 Expression을 t() 호출로 교체
                *expr = t_call;
            }
        }
        
        // 재귀적으로 자식 노드 방문
        expr.visit_mut_children_with(self);
    }
    
    /// StringLiteral 변환 (하위 호환성을 위해 유지)
    fn visit_mut_str(&mut self, n: &mut Str) {
        // visit_mut_expr에서 처리하므로 여기서는 아무것도 하지 않음
        let _ = n;
    }

    /// TemplateLiteral 변환
    /// TypeScript 버전과 동일한 로직:
    /// 1. i18n-ignore 주석이 있는 경우 스킵
    /// 2. 이미 t()로 래핑된 경우 스킵
    /// 3. 템플릿 리터럴의 모든 부분에 하나라도 한국어가 있는지 확인
    /// 4. 템플릿 리터럴을 i18next interpolation 형식으로 변환
    ///    예: `안녕 ${name}` → t(`안녕 {{name}}`, { name })
    fn visit_mut_tpl(&mut self, n: &mut Tpl) {
        // TODO: shouldSkipPath 및 hasIgnoreComment로 스킵 확인
        // TODO: 이미 t()로 래핑된 경우 스킵
        // TODO: 템플릿 리터럴의 모든 부분에 하나라도 한국어가 있는지 확인
        // TODO: Wtf8Atom을 문자열로 변환하는 올바른 방법 찾기
        // 현재는 소스코드에서 직접 검사
        if RegexPatterns::korean_text().is_match(&self.source_code) {
            self.was_modified = true;
            // TODO: 실제로는 i18next interpolation 형식으로 변환
            // 예: `안녕 ${name}` → t(`안녕 {{name}}`, { name })
        }
        let _ = n;
    }

    /// JSXText 변환
    /// TypeScript 버전과 동일한 로직:
    /// 1. i18n-ignore 주석이 있는 경우 스킵
    /// 2. 빈 텍스트나 공백만 있는 경우 스킵
    /// 3. 한국어가 포함된 텍스트만 처리
    /// 4. t() 함수 호출로 감싸기
    fn visit_mut_jsx_text(&mut self, n: &mut JSXText) {
        // TODO: hasIgnoreComment로 스킵 확인
        // TODO: Wtf8Atom을 문자열로 변환하는 올바른 방법 찾기
        // 현재는 소스코드에서 직접 검사
        if RegexPatterns::korean_text().is_match(&self.source_code) {
            self.was_modified = true;
            // TODO: 실제로는 JSXExpressionContainer로 감싸야 함
            // 현재는 플래그만 설정
        }
        let _ = n;
    }
}

/// Module을 변환하고 결과 반환
pub fn transform_module(module: &mut Module, source_code: String) -> TransformResult {
    let mut transformer = TranslationTransformer::new(source_code);
    module.visit_mut_with(&mut transformer);
    TransformResult::new(transformer.was_modified)
}
